pilot/pkg/bootstrap/config_compare.go

```diff
@@ -15,22 +15,39 @@
 package bootstrap
 
 import (
+	"reflect"
 	"strings"
 
 	gogoproto "github.com/gogo/protobuf/proto" // nolint: depguard
 	"google.golang.org/protobuf/proto"
+	gateway "sigs.k8s.io/gateway-api/apis/v1"
 
 	"istio.io/istio/pkg/config"
 )
 
 // needsPush checks whether the passed in config has same spec and hence push needs
 // to be triggered. This is to avoid unnecessary pushes only when labels have changed
 // for example.
-func needsPush(prev config.Config, curr config.Config) bool {
+func needsPush(prev config.Config, curr config.Config) (toPush bool, skipGenXDS bool) {
 	if prev.GroupVersionKind != curr.GroupVersionKind {
 		// This should never happen.
-		return true
+		return true, false
+	}
+
+	if prev.GroupVersionKind.Group == gateway.GroupName {
+		// We always push for Gateway API resources so that the status can be repaired.
+		return true, needsPushXDSForGatewayAPI(&prev, &curr)
 	}
+
+	return needsPushForOther(&prev, &curr), false
+}
+
+func needsPushXDSForGatewayAPI(prev *config.Config, curr *config.Config) bool {
```

Comment:
Oh I see what's happening; this name is misleading. The caller skips if this returns true but it's named incorrectly. Can you make the naming consistent?

Reply:
@keithmattix 
Thanks for your review. Updated.